File: mr-plugin/model-registry-ui/src/api/data_types/assetClass.ts
// Types for nested objects in Attributes
export interface AssetClassInterface {
	id: string;
	name: string;
	title: string;
	description: string;
	class_type: string;
	tags: string[];
	owner: string;
	created_at: string;
	create_by: string;
	modified_at: string;
	metadata: Record<string, any>;
	attributes: Record<string, any>;
	status: number;
	counter: number;
	[key: string]: unknown; // Index signature to handle dynamic properties
}

// AssetClass implementation
export class AssetClass implements AssetClassInterface {
	id: string;
	name: string;
	title: string;
	description: string;
	class_type: string;
	tags: string[];
	owner: string;
	created_at: string;
	create_by: string;
	modified_at: string;
	metadata: Record<string, any>;
	attributes: any;
	status: number;
	counter: number;
	[key: string]: unknown;
	
	constructor(data: Partial<AssetClassInterface>) {
		this.id = data.id || '';
		this.name = data.name || '';
		this.title = data.title || '';
		this.description = data.description || '';
		this.class_type = data.class_type || '';
		this.tags = data.tags || [];
		this.owner = data.owner || '';
		this.created_at = data.created_at || '';
		this.create_by = data.create_by || '';
		this.modified_at = data.modified_at || '';
		this.metadata = data.metadata || {};
		this.attributes = data.attributes || {};
		this.status = data.status || 1;
		this.counter = data.counter || 0;
	}
}

File: mr-plugin/server/mr-core/mr_core/models/asset_class.py
from peewee import *
from playhouse.postgres_ext import JSONField
from .base.base import BaseModel
from .project import Project


class AssetClass(BaseModel):
    project = ForeignKeyField(Project, backref='asset_classes', on_delete='CASCADE', null=False)
    name = CharField(unique=False)
    counter = IntegerField(default=0)
    owner = CharField(null=False)
    title = CharField(null=False, default="n/a")
    description = TextField(null=False, default="n/a")
    readme = TextField(null=True)
    class_type = CharField(null=False, default="n/a")
    attributes = JSONField(default=dict)  # attributes of the asset class
    metadata = JSONField(null=True, default=dict)  # additional metadata
    tags = JSONField(null=True, default=list)

    class Meta:
        indexes = (
            (('project', 'name'), True),
        )

    @classmethod
    def create(cls, user=None, **query):
        query["owner"] = query.get("owner", user)
        return super(AssetClass, cls).create(user, **query)

    @classmethod
    def serialize_fields(cls):
        return [
            "id",
            "name",
            "project",
            "created_at",
            "created_by",
            "owner",
            "title",
            "description",
            "class_type",
            "modified_at",
            "modified_by",
            "attributes",
            "status",
            "metadata"
        ]


